sisl/io/orca/output.py

```diff
@@ -0,0 +1,175 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+import numpy as np
+from .sile import SileORCA
+from ..sile import add_sile, sile_fh_open
+
+from sisl.utils import PropertyDict
+from sisl._internal import set_module
+from sisl import Geometry
+
+__all__ = ['outputSileORCA']
+
+
+@set_module("sisl.io.orca")
+class outputSileORCA(SileORCA):
+    """ Output file from ORCA """
+
+    @sile_fh_open()
+    def completed(self):
+        """ True if the full file has been read and "ORCA TERMINATED NORMALLY"" was found. """
+        return self.step_to("ORCA TERMINATED NORMALLY")[0]
+
+
+    @property
+    def _natoms(self):
+        f, line = self.step_to("Number of atoms")
+        v = line.split()
+        return int(v[-1])
+
+
+    def _read_atomic_block(self, natoms):
+        itt = iter(self)
+        next(itt) # skip ---
+        A = np.empty((natoms, 2), np.float64)
+        for ia in range(natoms):
+            line = next(itt)
+            v = line.split()
+            A[ia] = float(v[-2]), float(v[-1])
+        return A
+
+
+    @sile_fh_open()
+    def read_mulliken_atomic(self):
```

Comment:
As far as I can see Mulliken is not implemented in siesta out `read_charge`. Things also look a bit different in ORCA.

Alternatively, what about combining the atomic/orbital-versions of my functions into `read_mulliken(projection={'orbital', 'atom'}, orbital=None)` and `read_loewdin(projection={'orbital', 'atom'}, orbital=None)`?

The `orbital` keyword allows to select `s`, `p`, `d` (sums over orbitals and shells) or `pz`, `dz2` etc for specific ones.

Reply:
True that mulliken is not implemented in out, it should be ;)

I would still prefer a single routine, and using the projection argument would be ideal, that should probably also be added to the siesta-out file (currently only in mulliken).

Reply:
OK, so a single function here of the form `read_charge(name={'mulliken', 'loewdin'}, projection={'orbital', 'atom'}, orbital=None)`?

Reply:
Now, the charges quoted in ORCA are summed over spin. Therefore there is also the SPIN block. For clarity, would it make sense to introduce another function `read_spin(name={'mulliken', 'loewdin'}, projection={'orbital', 'atom'}, orbital=None)`?

Reply:
What the ORCA output contains are the CHARGE=up+down and SPIN=up-down blocks. One could of course disentangle this raw information to present CHARGE[up, down] to the user. Would you prefer this approach? The problem may be that the user is not getting processed data instead of the info directly from the output.

Reply:
I think as long as the documentation is clear what is returned, it should be sufficient. By guess would be that users are also interested in the spin-splitting of the charges, so summing/diff an axis should be trivial.